### ISSUE TITLE
Type is not narrowed in else clause following check via type()

### ISSUE THREAD

**Issue by eroller** [python/mypy#18315]
**Bug Report**

Type is not narrowed in else clause following check via type()

**To Reproduce**

```python
def check_type(a: int | str) -> None:
    if type(a) is int:
        my_int: int = a
    else:
        my_str: str = a # error: Incompatible types in assignment (expression has type "int | str", variable has type "str")  [assignment]
```
https://mypy-play.net/?mypy=latest&python=3.13&gist=20dd8e7b2f2f3a3795a78360b3627be7

**Expected Behavior**
`Success: no issues found in 1 source file`

**Actual Behavior**
`error: Incompatible types in assignment (expression has type "int | str", variable has type "str")  [assignment]`

**Your Environment**
- mypy 1.13.0
- python 3.13

**Comment by eroller** [comment]
a bug (#10457) on the if clause was previously fixed

**Comment by JelleZijlstra** [comment]
This behavior is correct. `a` might be an instance of a subtype of `int`.

**Comment by eroller** [comment]
Thank you. I guess that is why isinstance should be the preferred way to perform type narrowing. When I used custom classes with the `@final` decorator then the type is properly narrowed via type() as expected.
